src/Stream.elm

```diff
@@ -203,40 +203,28 @@ sample f varying events =
           (Native.Signal.initialValue varying, fromVarying varying)
 
       sampleEvents =
-          Native.Signal.genericMerge (\(Sample s) (Update u) -> SampleAndUpdate s u)
-            (map Sample events)
-            (map Update varyingUpdates)
+          Native.Signal.genericMerge (\s u -> (fst s, snd u))
+            (map (\s -> (Just s, Nothing)) events)
+            (map (\u -> (Nothing, Just u)) varyingUpdates)
   in
       fold sampleUpdate { state = initialValue, trigger = Nothing } sampleEvents
         |> fromVarying
         |> filterMap (\state -> Maybe.map (f state.state) state.trigger)
 
 
-type SampleEvent a b = Sample a | Update b | SampleAndUpdate a b
+type alias SampleEvent b a = (Maybe b, Maybe a)
 
 
-type alias SampleState a b =
-    { state : b
-    , trigger : Maybe a
+type alias SampleState b a =
+    { state : a
+    , trigger : Maybe b
     }
 
 
-sampleUpdate : SampleEvent a b -> SampleState a b -> SampleState a b
-sampleUpdate event state =
-  case event of
-    Sample a ->
-        { state = state.state
-        , trigger = Just a
-        }
-
-    Update b ->
-        { state = b
-        , trigger = Nothing
-        }
-
-    SampleAndUpdate a b ->
-        { state = b
-        , trigger = Just a
+sampleUpdate : SampleEvent b a -> SampleState b a -> SampleState b a
+sampleUpdate (ms,mu) state =
```

Comment:
Let's do better names than `mu` and `ms`. I don't know what they mean, and saving characters is not a good goal.